Support voltage formulas

shsms opened the pull request:
This is also a major internal revamp of how metrics are represented and how the component graph is accessed, but the external interface only has minor changes.

shsms committed:
Upgrade to latest component-graph

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Rename `Formula` → `AggregationFormula`

Because there will be another type of formula that doesn't support
aggregation.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add new Metric representation bound to a formula type

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Move `AggregationFormula` into its own module

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Defines a `Formula` trait and implement it for `AggregationFormula`

This will be implemented by the upcoming `CoalesceFormula` as well.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Implement a generic way to create formulas from the component graph

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Upgrade the logical meter to use the new metric types

This also enables us to delegate the component graph formula creation
to the logical meter formulas, through the `GraphFormulaProvider`
trait.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Remove old Metric definitions

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Implement `CoalesceFormula`

And introduce support for voltage and frequency metrics using
`CoalesceFormula`s.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Start a resampler for each requested metric for a component

This is a bugfix, because earlier there was only one resampler for
each component and data from all metrics were coming from that one
resampler, which had the data only for the first metric that was
subscribed to.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update example to also stream grid voltage

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
Draft because it depends on unmerged code from the component graph repo.

shsms commented:
And based on #6

shsms commented:
Ready for review now.

niklas-timpe commented on Cargo.toml:
```diff
@@ -9,7 +9,8 @@ path = "src/lib.rs"
 [dependencies]
 chrono = "0.4"
-frequenz-microgrid-component-graph = { git = "https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs.git", rev = "ab3b998" }
+# frequenz-microgrid-component-graph = { git = "https://github.com/shsms/frequenz-microgrid-component-graph-rs.git", branch = "coalesce-formulas" }
```
should that stay here?

shsms replied:
oops, no

niklas-timpe commented on src/logical_meter/metric.rs:
```diff
@@ -3,8 +3,46 @@
+    };
+}
+
+define_metric! {
```
Currently Girdguard is having a wrapper over the old Metric. This will be not as easy anymore with this approach.
We could remove the wrapper entirely though by adding the following traits:
```
#[derive(Debug, Clone, Copy, PartialEq, Eq)]

#and

impl Hash for ChannelMetric {
    fn hash<H: Hasher>(&self, state: &mut H) {
        // Just hash the discriminant or some identifier
        (self.0 as i32).hash(state);
    }
}
```

niklas-timpe replied:
no need anymore

niklas-timpe commented on src/logical_meter/logical_meter_handle.rs:
```diff
@@ -56,154 +56,98 @@ impl LogicalMeterHandle {
+        M::FormulaType::producer(&self.graph, metric, self.instructions_tx.clone())
     }
 
     pub fn coalesce(
```
Why not updating coalesce as well?
The metric used here is directly from the microgrid api?

shsms replied:
I will revamp the coalesce mechanism in upcoming PRs, this will go away then.

shsms committed:
Fix doc-comment typo

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Rename stream status channel for clarity

The channel used to communicate the status of a component data stream
was called `stream_stopped_tx/rx`, which is no longer accurate.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Group resampled data by metrics and then by component ID

Earlier we were picking a random metric and sending its data to all
formulas, which was incorrect.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

niklas-timpe reviewed (approved):
lgtm
